### DISCUSSION TITLE
How to run tests for just 1 test case?

### DISCUSSION THREAD

**Discussion by simkimsia** [PyGithub/PyGithub#2013] [Q&A]
I forked the repo to add more features that allow me to fetch template repos for authenticateduser.

I wrote this

```
def testGetTemplateRepos(self):
        self.assertListKeyEqual(
        ..
```

under tests/AuthenticatedUser.py

How do I run the test for this?

**Comment by s-t-e-v-e-n-k** [comment]
I use `tox` to run the test suite, for example: `tox -epy38` -- this also allows you to only run matching tests. From `pytest` documentation: https://docs.pytest.org/en/6.2.x/usage.html#specifying-tests-selecting-tests , so you can do `tox -epy38 -- -k '<selection criteria>'`

**Reply by xmo-odoo** [reply]
Running tests by node ids is also very useful, because while it's a bit longer and more complicated than `-k` it is much more precise, `-k` will basically check if a given class or method name contains the needle and run those that do, so if some test names are subsets of others (e.g. `testCreateGitTree` and `testCreateGitTreeWithBaseTree` in pygithub) there will be no way to run the first without running the second.

The only annoyance with nodeids is pytest doesn't really want to hand them out without running the tests: you need `pytest --collect-only -qqq` for pytest to switch to dumping nodeids instead of the pseudo-xml it uses by default.

**Comment by s-t-e-v-e-n-k** [comment]
`-k` is actually incredibly powerful because you can build up expressions that all have to be true for it match a test.

```
% tox -epy38 -- -k 'testAttributes' --co | grep collected
collected 723 items / 670 deselected / 53 selected
=============== 53/723 tests collected (670 deselected) in 1.89s ===============
% tox -epy38 -- -k 'testAttributes and Repository' --co | grep collected
collected 723 items / 721 deselected / 2 selected
=============== 2/723 tests collected (721 deselected) in 1.87s ================
% tox -epy38 -- -k 'testAttributes and Repository and not Key' --co | grep collected
collected 723 items / 722 deselected / 1 selected
=============== 1/723 tests collected (722 deselected) in 1.90s ================
```

As you can see, you can easily work your way down to just the single testcase you want, even if it's an incredibly common testcase name -- I think testAttributes is the most comment one we have.
